File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int startingPrice = prices[0];
        int highestPrice = 0;

        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > startingPrice) {
                highestPrice = Math.max(highestPrice, prices[i] - startingPrice);
            } else {
                startingPrice = prices[i];
            }
        }
        return highestPrice;

    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        min_price = prices[0]
        best = 0

        for i in range(n):
            min_price = min(min_price, prices[i])
            best = max(best, prices[i] - min_price)
        return best
